### PR TITLE
Refactor reloadable turrets

### PR DESCRIPTION
**Pull request by perkinslr**
## Additions

Allow other mods' turrets to use CE gizmos and reloading logic via a reloadable `TurretRegistry`

## Changes

Most usages of `Building_TurretGunCE` have been replaced with the generic `Building_Turret`.  
An extension class `CombatExtended.Compatibility.TurretRegistry` provides access to the parts of `Building_TurretGunCE` which are actually needed elsewhere in the code (getters for `Gun`, `CompAmmo`, `CompMannable`, getter and setter for `isReloading`).  
A public `RegisterTurret` method allows other mods to register subclasses of `Building_Turret`, for use with those same getters and setters.

## Reasoning

Combat Extended uses a subclass of `Building_TurretGun`, which precludes sharing code with other mods subclassing `Building_Turret`.  For mods which use CE gun definitions, this is particularly problematic, as they have to reach into CE internals to properly reload guns contained in their turrets.  At best, this leads to methods copied from CE, with the typenames changed.  In practice, mods stay broken, or the copied code slips out of sync.  Moving turret reloading to a modular system should make it relatively easy to add support for other turrets, either from within the CE codebase, or for other mod authors to do it from their code.

## Alternatives

Continue to fix issues one at a time, using reflection and dealing with breakages whenever internal structures change.

## Testing

Check tests you have performed:
- [X] Compiles without warnings
- [X] Game runs without errors
- [X] (For compatibility patches) ...with and without patched mod loaded
- [X] Playtested a colony (minutes)

## Other Notes

When there are no mods using the new API, it should get JITted out, so the performance and stability impact *should* be quite minimal.  That said, this is a fairly involved set of changes, so we should make sure to test extensively before we push this live.

### CODE REVIEW

File: Source/CombatExtended/Compatibility/TurretRegistry.cs [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+		return ceturret.MannableComp;
+	    }
+	    return turret.TryGetComp<CompMannable>();
+	}
+
+	public static bool GetReloadable(this Building_Turret turret) {
+	    return turret.GetAmmo()?.HasMagazine ?? false;
+	}
+	
+	public static bool ShouldReload(this Building_Turret turret, float threshold=0.5f, bool ensure_ammo_type=true) {
```
Comment (simplyWiri) [inline Source/CombatExtended/Compatibility/TurretRegistry.cs:132]: Is snake case used for default parameters ? I do not recall seeing this in other parts of the code

File: Source/CombatExtended/Compatibility/TurretRegistry.cs [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+		if (getGun.TryGetValue(turret.GetType(), out gfunc)) {
+		    return gfunc(turret);
+		}
+		
+	    }
+	    CELogger.Warn("Asked to get gun on an unknown turret type: "+turret);
+	    return null;
+	}
+
+	public static CompMannable GetMannable(this Building_Turret turret) {
```
Comment (simplyWiri) [inline Source/CombatExtended/Compatibility/TurretRegistry.cs:120]:
```csharp
public static CompMannable GetMannable(this Building_Turret turret) {
            if(turret is Building_TurretGunCE ceturret) return ceturret.MannableComp;
	    return turret.TryGetComp<CompMannable>();
}
```

File: Source/CombatExtended/CombatExtended/Jobs/WorkGiver_ReloadTurret.cs [hunk 3]
```diff
@@ -61,20 +65,21 @@ public override bool ShouldSkip(Pawn pawn, bool forced = false)
         public override bool HasJobOnThing(Pawn pawn, Thing t, bool forced = false)
         {
             // Should never happen anymore, as only TurretGunCEs should be returned from PotentialWorkThingsGlobal
-            if (!(t is Building_TurretGunCE)) {
+            Building_Turret turret = t as Building_Turret;
+            if (turret == null) {
```
Comment (simplyWiri) [inline Source/CombatExtended/CombatExtended/Jobs/WorkGiver_ReloadTurret.cs:69]: Is there a reason for this check ? The only items stored in the turret tracker mapcomp are turrets of the same type.

**Review by simplyWiri — APPROVED** [review]
There are quite a few uses of `as` followed by a null check & branching which could be replaced with `is`. However didn't notice anything which could block a merge - happy to go along with this unless there are changes still in the works or someone else mentions issues.

I have not play tested this to ensure the desired changes are made - just looked over the code

File: Source/CombatExtended/Compatibility/TurretRegistry.cs (hunk 1 above)
Comment (perkinslr) [reply Source/CombatExtended/Compatibility/TurretRegistry.cs:132]: No, that's just me being sloppy after being away fro mthe codebase for too long.

Comment (perkinslr) [reply Source/CombatExtended/Compatibility/TurretRegistry.cs:132]: Also, that parameter isn't used...

File: Source/CombatExtended/Compatibility/TurretRegistry.cs (hunk 2 above)
Comment (perkinslr) [reply Source/CombatExtended/Compatibility/TurretRegistry.cs:120]: At the time I first wrote this, I was trying to keep to mono-compiler C#7-compatible language features.  There's probably a few other places that can see a similar benefit scattered throughout the code base.

File: Source/CombatExtended/CombatExtended/Jobs/WorkGiver_ReloadTurret.cs (hunk 3 above)
Comment (perkinslr) [reply Source/CombatExtended/CombatExtended/Jobs/WorkGiver_ReloadTurret.cs:69]: I believe it is historical, given the comment on line 67.  I have updated the comment and condensed the cast+null-check.  We have to do the cast anyway, as the supercall's HasJobOnThing specifies `Thing` as the incoming type.  If the null check really is redundant, I'm hopeful that the JIT will ellide it (I believe it should).

**Review by simplyWiri — APPROVED** [review]

**Review by N7Huntsman — APPROVED** [review]

**Pull request merged**